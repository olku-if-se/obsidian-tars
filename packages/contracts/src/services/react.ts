/**
 * React Bridge service interface for managing React components in Obsidian
 */

import type { App } from 'obsidian'

// Generic React component type to avoid direct React dependency
export type ReactComponentType<T extends Record<string, unknown> = Record<string, unknown>> = (props: T) => unknown

export interface IReactBridge {
	/**
	 * Mount a React component into a container element
	 */
	mount<T extends Record<string, unknown>>(container: HTMLElement, component: ReactComponentType<T>, props: T): void

	/**
	 * Unmount a React component from a container element
	 */
	unmount(container: HTMLElement): void

	/**
	 * Unmount all mounted React components
	 */
	unmountAll(): void

	/**
	 * Check if a container has a mounted React component
	 */
	hasRoot(container: HTMLElement): boolean

	/**
	 * Get the number of currently mounted components
	 */
	getMountedCount(): number
}

export interface IReactBridgeManager {
	/**
	 * Get the React bridge instance
	 */
	getReactBridge(): IReactBridge

	/**
	 * Initialize the React bridge with Obsidian app
	 */
	initialize(app: App): void

	/**
	 * Clean up all React components
	 */
	dispose(): void

	/**
	 * Check if React bridge is ready
	 */
	isReady(): boolean
}
