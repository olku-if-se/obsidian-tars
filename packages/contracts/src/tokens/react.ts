export const ReactBridgeToken = Symbol('ReactBridge')
export const ReactBridgeManagerToken = Symbol('ReactBridgeManager')
export const StatusBarElementToken = Symbol('StatusBarElement')

export const AppToken = Symbol('App')
export const TarsPluginToken = Symbol('TarsPlugin')

// Type assertions to help TypeScript understand the token-service relationship
export type ReactBridgeToken = typeof ReactBridgeToken
export type ReactBridgeManagerToken = typeof ReactBridgeManagerToken
export type StatusBarElementToken = typeof StatusBarElementToken
export type AppToken = typeof AppToken
export type TarsPluginToken = typeof TarsPluginToken
