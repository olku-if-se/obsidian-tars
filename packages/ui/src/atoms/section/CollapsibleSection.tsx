import clsx from 'clsx'
import type React from 'react'
import { forwardRef, useCallback, useState } from 'react'
import styles from './CollapsibleSection.module.css'

interface CollapsibleSectionProps
  extends Omit<React.ComponentPropsWithoutRef<'details'>, 'title' | 'onToggle'> {
	// Heading content for the summary; HTML 'title' is omitted above
	title: React.ReactNode
	defaultOpen?: boolean
	// Library-level toggle callback (not native Details onToggle)
	onToggle?: (open: boolean) => void | Promise<void>
}

export const CollapsibleSection = forwardRef<HTMLDetailsElement, CollapsibleSectionProps>(
	({ title, open, defaultOpen = false, onToggle, children, className, ...rest }, ref) => {
		const [uncontrolledOpen, setUncontrolledOpen] = useState(defaultOpen)
		const isControlled = open !== undefined
		const isOpen = isControlled ? (open as boolean) : uncontrolledOpen

		// Sync with native <details> toggle; read truth from the element
		const handleDetailsToggle = useCallback(
			(e: React.SyntheticEvent<HTMLDetailsElement>) => {
				const next = e.currentTarget.open
				if (!isControlled) setUncontrolledOpen(next)
				onToggle?.(next)
			},
			[isControlled, onToggle]
		)

		return (
			<details
				ref={ref}
				className={clsx(styles.collapsible, className)}
				open={isOpen}
				onToggle={handleDetailsToggle}
				{...rest}
			>
				<summary className={styles.summary}>
					<span className={styles.summaryText}>{title}</span>
					<span className={clsx(styles.arrow, isOpen && styles.arrowOpen)} aria-hidden="true">
						â–¼
					</span>
				</summary>
				<div className={styles.content}>{children}</div>
			</details>
		)
	}
)

CollapsibleSection.displayName = 'CollapsibleSection'
