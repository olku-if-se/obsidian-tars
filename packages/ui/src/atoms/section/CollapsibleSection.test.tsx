import { render, screen, fireEvent } from '@testing-library/react'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import { CollapsibleSection } from './CollapsibleSection'

describe('CollapsibleSection', () => {
	beforeEach(() => {
		vi.clearAllMocks()
	})

	it('renders title correctly', () => {
		render(
			<CollapsibleSection title='Test Section'>
				<p>Content</p>
			</CollapsibleSection>
		)

		expect(screen.getByText('Test Section')).toBeInTheDocument()
	})

	it('renders children when open', () => {
		render(
			<CollapsibleSection title='Test Section' defaultOpen>
				<p>Content</p>
			</CollapsibleSection>
		)

		expect(screen.getByText('Content')).toBeInTheDocument()
	})

	it('does not render children when closed', () => {
		render(
			<CollapsibleSection title='Test Section' defaultOpen={false}>
				<p>Content</p>
			</CollapsibleSection>
		)

		expect(screen.queryByText('Content')).not.toBeInTheDocument()
	})

	it('handles controlled state correctly', () => {
		const handleToggle = vi.fn()
		render(
			<CollapsibleSection title='Test Section' open={false} onToggle={handleToggle}>
				<p>Content</p>
			</CollapsibleSection>
		)

		const summary = screen.getByText('Test Section').closest('summary')
		fireEvent.click(summary!)

		expect(handleToggle).toHaveBeenCalledWith(true)
	})

	it('handles uncontrolled state correctly', () => {
		render(
			<CollapsibleSection title='Test Section' defaultOpen={false}>
				<p>Content</p>
			</CollapsibleSection>
		)

		const summary = screen.getByText('Test Section').closest('summary')
		fireEvent.click(summary!)

		expect(screen.getByText('Content')).toBeInTheDocument()
	})

	it('calls onToggle when toggled', async () => {
		const handleToggle = vi.fn()
		render(
			<CollapsibleSection title='Test Section' onToggle={handleToggle}>
				<p>Content</p>
			</CollapsibleSection>
		)

		const summary = screen.getByText('Test Section').closest('summary')
		fireEvent.click(summary!)

		expect(handleToggle).toHaveBeenCalledWith(true)
	})

	it('handles async onToggle without errors', async () => {
		const handleToggle = vi.fn().mockResolvedValue(undefined)
		const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {})

		render(
			<CollapsibleSection title='Test Section' onToggle={handleToggle}>
				<p>Content</p>
			</CollapsibleSection>
		)

		const summary = screen.getByText('Test Section').closest('summary')
		fireEvent.click(summary!)

		// Wait for async operation
		await vi.waitFor(() => expect(handleToggle).toHaveBeenCalled())

		expect(consoleSpy).not.toHaveBeenCalled()
		consoleSpy.mockRestore()
	})

	it('warns about controlled/uncontrolled mismatch in development', () => {
		const originalEnv = process.env.NODE_ENV
		process.env.NODE_ENV = 'development'
		const consoleSpy = vi.spyOn(console, 'warn').mockImplementation(() => {})

		const { rerender } = render(
			<CollapsibleSection title='Test Section'>
				<p>Content</p>
			</CollapsibleSection>
		)

		// Switch to controlled
		rerender(
			<CollapsibleSection title='Test Section' open={true}>
				<p>Content</p>
			</CollapsibleSection>
		)

		expect(consoleSpy).toHaveBeenCalledWith(
			expect.stringContaining('changing from uncontrolled to controlled')
		)

		consoleSpy.mockRestore()
		process.env.NODE_ENV = originalEnv
	})

	it('applies custom className correctly', () => {
		render(
			<CollapsibleSection title='Test Section' className='custom-class'>
				<p>Content</p>
			</CollapsibleSection>
		)

		const details = screen.getByRole('group')
		expect(details).toHaveClass('custom-class')
	})

	it('has correct accessibility attributes', () => {
		render(
			<CollapsibleSection title='Test Section' open>
				<p>Content</p>
			</CollapsibleSection>
		)

		const details = screen.getByRole('group')
		expect(details).toHaveAttribute('open')
		expect(details).toHaveAttribute('data-state', 'open')

		const arrow = screen.getByText('â–¼')
		expect(arrow).toHaveAttribute('aria-hidden', 'true')
		expect(arrow).toHaveAttribute('data-open', 'true')
	})

	it('matches snapshot', () => {
		const { asFragment } = render(
			<CollapsibleSection title='Test Section' defaultOpen>
				<p>Content</p>
			</CollapsibleSection>
		)

		expect(asFragment()).toMatchSnapshot()
	})
})