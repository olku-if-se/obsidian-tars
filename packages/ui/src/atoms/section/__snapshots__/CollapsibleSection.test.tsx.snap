// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CollapsibleSection > matches snapshot 1`] = `
<DocumentFragment>
  <details
    class="_collapsible_9f9079"
    data-state="open"
    open=""
  >
    <summary
      class="_summary_9f9079"
    >
      <span
        class="_summaryText_9f9079"
      >
        Test Section
      </span>
      <span
        aria-hidden="true"
        class="_arrow_9f9079 _arrowOpen_9f9079"
        data-open="true"
      >
        â–¼
      </span>
    </summary>
    <div
      class="_content_9f9079"
      data-hidden="false"
    >
      <p>
        Content
      </p>
    </div>
  </details>
</DocumentFragment>
`;
